src/main/java/org/wildfly/security/auth/callback/MechanismInformationCallback.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2016 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.auth.callback;
+
+import java.io.Serializable;
+
+import javax.security.auth.callback.CallbackHandler;
+
+/**
+ * A {@link Callback} to pass the information about the current mechanism to the {@link CallbackHandler}.
+ *
+ * As an informational {@code Callback} it is optional for the {@code CallbackHandler} to handle this.
+ *
+ * @author <a href="mailto:darran.lofthouse@jboss.com">Darran Lofthouse</a>
+ */
+public class MechanismInformationCallback implements ExtendedCallback, Serializable {
+
+    private static final long serialVersionUID = -8376970360160709801L;
+
+    private final String mechanismType;
+    private final String mechanismName;
+    private final String hostName;
+    private final String protocol;
+
+    /**
+     * Construct a new instance with the appropriate mechanism information.
+     *
+     * @param mechanismType the mechanism type.
+     * @param mechanismName the name of the mechanism.
+     * @param hostName the host name for the current authentication.
+     * @param protocol the protocol for the current authentication.
+     */
+    public MechanismInformationCallback(final String mechanismType, final String mechanismName, final String hostName, final String protocol) {
```

Comment:
Yeah I think default methods will have to cover it - whatever we come up backwards compatibility means all new information is optional so will effectively have a default value - default methods in the interface cover this.